File: src/routes/api/ws/exec.ts
import net from 'node:net'
import { createFileRoute } from '@tanstack/react-router'
import { defineHooks } from 'crossws'
import type Docker from 'dockerode'
import { dockerClient } from '#/lib/docker/client'

type ExecSession = {
  socket: net.Socket
  exec: Docker.Exec
}

const sessions = new Map<string, ExecSession>()

function startExecHijack(
  execId: string,
): Promise<{ socket: net.Socket; remaining: Buffer }> {
  return new Promise((resolve, reject) => {
    const socket = net.createConnection('/var/run/docker.sock')
    const body = JSON.stringify({ Detach: false, Tty: true })
    const request = [
      `POST /exec/${execId}/start HTTP/1.1`,
      'Host: localhost',
      `Content-Length: ${Buffer.byteLength(body)}`,
      'Content-Type: application/json',
      'Upgrade: tcp',
      'Connection: Upgrade',
      '',
      body,
    ].join('\r\n')

    let headerBuf = Buffer.alloc(0)
    let upgraded = false

    const onData = (chunk: Buffer) => {
      if (upgraded) return
      headerBuf = Buffer.concat([headerBuf, chunk])
      const idx = headerBuf.indexOf('\r\n\r\n')
      if (idx !== -1) {
        upgraded = true
        socket.removeListener('data', onData)
        const statusLine = headerBuf.slice(0, idx).toString().split('\r\n')[0]
        if (statusLine.includes('101') || statusLine.includes('200')) {
          resolve({ socket, remaining: headerBuf.slice(idx + 4) })
        } else {
          reject(new Error(`exec start failed: ${statusLine}`))
          socket.destroy()
        }
      }
    }

    socket.on('data', onData)
    socket.on('error', reject)
    socket.once('connect', () => socket.write(request))
  })
}

const hooks = defineHooks({
  async message(peer, message) {
    const raw = message.rawData

    // Binary messages → raw terminal input
    if (
      raw instanceof Uint8Array ||
      raw instanceof ArrayBuffer ||
      Buffer.isBuffer(raw)
    ) {
      const session = sessions.get(peer.id)
      if (session)
        session.socket.write(
          Buffer.isBuffer(raw) ? raw : Buffer.from(raw as ArrayBuffer),
        )
      return
    }

    try {
      const msg = message.json()

      if (msg.type === 'init') {
        const containerId = msg.containerId as string
        const cols = (msg.cols as number) ?? 80
        const rows = (msg.rows as number) ?? 24
        const shell = (msg.shell as string) ?? '/bin/sh'

        try {
          const container = dockerClient.getContainer(containerId)
          const exec = await container.exec({
            Cmd: [shell],
            AttachStdin: true,
            AttachStdout: true,
            AttachStderr: true,
            Tty: true,
            Env: [`TERM=xterm-256color`, `COLUMNS=${cols}`, `LINES=${rows}`],
          })

          const { socket, remaining } = await startExecHijack(exec.id)

          try {
            await exec.resize({ h: rows, w: cols })
          } catch {
            // process may have already exited (shell not found)
          }

          sessions.set(peer.id, { socket, exec })

          if (remaining.length > 0) peer.send(remaining)

          socket.on('data', (chunk: Buffer) => peer.send(chunk))

          socket.on('end', () => {
            peer.send(JSON.stringify({ type: 'exit' }))
            sessions.delete(peer.id)
          })

          socket.on('error', (err: Error) => {
            peer.send(JSON.stringify({ type: 'error', message: err.message }))
            sessions.delete(peer.id)
          })
        } catch (err) {
          peer.send(JSON.stringify({ type: 'error', message: String(err) }))
        }

        return
      }

      if (msg.type === 'resize') {
        const session = sessions.get(peer.id)
        if (session) {
          await session.exec.resize({
            h: msg.rows as number,
            w: msg.cols as number,
          })
        }
        return
      }
    } catch {
      // Not JSON — raw terminal input
      const session = sessions.get(peer.id)
      if (session) session.socket.write(message.text())
      return
    }
  },

  close(peer) {
    const session = sessions.get(peer.id)
    if (session) {
      session.socket.destroy()
      sessions.delete(peer.id)
    }
  },
})

export const Route = createFileRoute('/api/ws/exec')({
  server: {
    handlers: {
      GET: async () => {
        return Object.assign(
          new Response('WebSocket upgrade is required.', {
            status: 426,
          }),
          {
            crossws: hooks,
          },
        )
      },
    },
  },
})

File: internal/server/api/ws/exec.go
package ws

import (
	"context"
	"encoding/json"
	"net"
	"net/http"
	"strconv"
	"sync"

	"github.com/coder/websocket"
	"github.com/docker/docker/api/types/container"
)

var authorizedShells = map[string]bool{
	"/bin/bash": true,
	"/bin/sh":   true,
	"/bin/zsh":  true,
	"/bin/fish": true,
}

func (d *Deps) handleExecWS(w http.ResponseWriter, r *http.Request) {
	conn, err := acceptWS(w, r)
	if err != nil {
		return
	}
	defer func() { _ = conn.CloseNow() }()

	ctx := r.Context()

	var (
		mu      sync.Mutex
		execID  string
		hijack  net.Conn
		cancel  context.CancelFunc
		started bool
	)

	for {
		typ, data, err := conn.Read(ctx)
		if err != nil {
			mu.Lock()
			if hijack != nil {
				if tc, ok := hijack.(interface{ CloseWrite() error }); ok {
					_ = tc.CloseWrite()
				}
			}
			mu.Unlock()
			if cancel != nil {
				cancel()
			}
			return
		}

		switch typ {
		case websocket.MessageBinary:
			mu.Lock()
			if hijack != nil && len(data) > 0 {
				_, _ = hijack.Write(data)
			}
			mu.Unlock()
			continue
		case websocket.MessageText:
			var msg struct {
				Type string `json:"type"`
			}
			if json.Unmarshal(data, &msg) != nil {
				mu.Lock()
				if hijack != nil {
					_, _ = hijack.Write(data)
				}
				mu.Unlock()
				continue
			}
			switch msg.Type {
			case "init":
				var init struct {
					Type        string `json:"type"`
					ContainerID string `json:"containerId"`
					Cols        int    `json:"cols"`
					Rows        int    `json:"rows"`
					Shell       string `json:"shell"`
				}
				if json.Unmarshal(data, &init) != nil {
					continue
				}
				if !authorizedShells[init.Shell] {
					_ = sendWSJSON(conn, map[string]string{"type": "error", "message": "shell not authorized"})
					continue
				}
				cols, rows := init.Cols, init.Rows
				if cols <= 0 {
					cols = 80
				}
				if rows <= 0 {
					rows = 24
				}
				execCtx, execCancel := context.WithCancel(ctx)
				cancel = execCancel
				hj, exec, err := d.startExec(execCtx, init.ContainerID, init.Shell, cols, rows)
				if err != nil {
					_ = sendWSJSON(conn, map[string]string{"type": "error", "message": err.Error()})
					execCancel()
					continue
				}
				mu.Lock()
				hijack = hj
				execID = exec
				started = true
				mu.Unlock()
				go d.pumpExec(conn, hj, ctx)
			case "resize":
				if !started {
					continue
				}
				var resize struct {
					Type string `json:"type"`
					Cols int    `json:"cols"`
					Rows int    `json:"rows"`
				}
				if json.Unmarshal(data, &resize) != nil {
					continue
				}
				if resize.Cols > 0 && resize.Rows > 0 {
					_ = d.Docker.ResizeExec(ctx, execID, resize.Rows, resize.Cols)
				}
			case "close":
				mu.Lock()
				if hijack != nil {
					if tc, ok := hijack.(interface{ CloseWrite() error }); ok {
						_ = tc.CloseWrite()
					}
				}
				mu.Unlock()
				if cancel != nil {
					cancel()
				}
				return
			default:
				mu.Lock()
				if hijack != nil {
					_, _ = hijack.Write(data)
				}
				mu.Unlock()
			}
		}
	}
}

func (d *Deps) pumpExec(conn *websocket.Conn, hijack net.Conn, ctx context.Context) {
	buf := make([]byte, 32*1024)
	for {
		n, err := hijack.Read(buf)
		if n > 0 {
			if werr := conn.Write(ctx, websocket.MessageBinary, buf[:n]); werr != nil {
				return
			}
		}
		if err != nil {
			_ = sendWSJSON(conn, map[string]string{"type": "exit"})
			return
		}
	}
}

func (d *Deps) startExec(ctx context.Context, containerID, shell string, cols, rows int) (net.Conn, string, error) {
	env := []string{
		"TERM=xterm-256color",
		"COLUMNS=" + strconv.Itoa(cols),
		"LINES=" + strconv.Itoa(rows),
	}
	createResp, err := d.Docker.CreateExec(ctx, containerID, container.ExecOptions{
		Cmd:          []string{shell},
		AttachStdin:  true,
		AttachStdout: true,
		AttachStderr: true,
		Tty:          true,
		Env:          env,
	})
	if err != nil {
		return nil, "", err
	}
	execID := createResp.ID
	hj, err := d.Docker.AttachExec(ctx, execID)
	if err != nil {
		return nil, "", err
	}
	_ = d.Docker.ResizeExec(ctx, execID, rows, cols)
	return hj, execID, nil
}
